source/Calamari/Deployment/PackageRetention/Model/PackageJournal.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System;
+using System.Linq;
+using Calamari.Common.Features.Processes.Semaphores;
+using Calamari.Common.Plumbing.Deployment.PackageRetention;
+using Calamari.Common.Plumbing.FileSystem;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Deployment.PackageRetention.Caching;
+using Calamari.Deployment.PackageRetention.Repositories;
+
+namespace Calamari.Deployment.PackageRetention.Model
+{
+    public class PackageJournal : IManagePackageCache
+    {
+        readonly IJournalRepository journalRepository;
+        readonly IRetentionAlgorithm retentionAlgorithm;
+        readonly ILog log;
+        readonly ICalamariFileSystem fileSystem;
+        readonly ISemaphoreFactory semaphoreFactory;
+
+        public PackageJournal(IJournalRepository journalRepository,
+                              ILog log,
+                              ICalamariFileSystem fileSystem,
+                              IRetentionAlgorithm retentionAlgorithm,
+                              ISemaphoreFactory semaphoreFactory)
+        {
+            this.journalRepository = journalRepository;
+            this.log = log;
+            this.fileSystem = fileSystem;
+            this.retentionAlgorithm = retentionAlgorithm;
+            this.semaphoreFactory = semaphoreFactory;
+        }
+
+        public void RegisterPackageUse(PackageIdentity package, ServerTaskId deploymentTaskId, ulong packageSizeBytes)
+        {
+            try
+            {
+                using (AcquireSemaphore())
```

Comment:
I know it's never really going to be an issue, but I did like the idea of the repository handling the semaphore - if we were to change to (say) using a SQL database-backed repository that handles concurrent operations ok, the semaphore wouldn't be needed.
Bit of a [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) situation though 😀

Reply:
Where would you put the semaphores in the repository to provide thread-safety? I think to ensure operations are atomic, the semaphore needs to span from the Load() to Commit().

Reply:
The pattern I used was to have the semaphore creation and repo `Load()` occur within the constructor (which was called using the factory `CreateJournalRepository()` method), and the semaphore closed in the `Dispose()` method .
That way the semaphore is transparent, like here: https://github.com/OctopusDeploy/Calamari/blob/e2d4794bda1b919fc1cea84b1b6463d853a28d85/source/Calamari/Deployment/PackageRetention/Model/Journal.cs#L78 and you don't have to worry about calling `Load()` (I guess similarly to how a repo over a DB might work).

Reply:
BTW don't feel like you have to change it or anything - it was more a point of interest than anything else 😃